src/Mvc/Mvc.Core/src/ModelBinding/Binders/DictionaryModelBinder.cs

```diff
@@ -124,20 +124,16 @@ public override async Task BindModelAsync(ModelBindingContext bindingContext)
         }
 
         await base.BindModelAsync(bindingContext);
-        if (!bindingContext.Result.IsModelSet)
-        {
-            // No match for the prefix at all.
-            return;
-        }
-
         var result = bindingContext.Result;
 
-        Debug.Assert(result.Model != null);
-        var model = (IDictionary<TKey, TValue?>)result.Model;
-        if (model.Count != 0)
+        if (result.IsModelSet)
         {
-            // ICollection<KeyValuePair<TKey, TValue>> approach was successful.
-            return;
+            Debug.Assert(result.Model != null);
+            if (result.Model is IDictionary<TKey, TValue?> { Count: > 0 })
```

Comment:
Confirming my understanding here: this addresses the scenario where the model has been bound successfully and the user _did not_ provide `null` as a default value?

Reply:
No, this is the success path, when we are able to bind.

Reply:
Yeah `model has been bound successfully` but in this case does not matter if it has a default value.  Of course, the default value will affect if the IsModelSet check before.